src/Utils/ProtoHelpers.php

```diff
@@ -78,31 +79,81 @@ public static function isMap(ProtoCatalog $catalog, FieldDescriptor $desc): bool
+            ])
+            ->toArray();
+    }
+
+    public static function routingBasedHeaderParams(ProtoCatalog $catalog, RoutingRule $routingRule, DescriptorProto $msg): array
     {
-        return Vector::new($routingRule->getRoutingParameters())
-            ->groupBy(
-                // Key: The field (or field chain) referenced by the RoutingParamter.
-                fn ($x) => $x->getField(),
-                // Value: The routing header config the RoutingParameter defines.
-                fn ($x) => [
-                    'getter' => static::buildGetterChain($catalog, $msg, $x->getField()),
-                    'key' => static::fieldOrTemplateVariable($x),
-                    'regex' => static::compileRoutingRegex($x),
-                    'root' => explode('.', $x->getField())[0],
-                ],
+
+        // Organzie rules by resulting key name.
+        $rulesByKey = Vector::new($routingRule->getRoutingParameters())
+            ->groupBy(fn($x) => static::fieldOrTemplateVariable($x));
+
+        // Organize "compiled" segment matchers by resulting key name.
+        $matchersByKey = $rulesByKey->mapValues(
+            fn($key, $routingRules) =>
+                $routingRules
+                    ->filter(fn($rule) => static::hasMatcher($rule))
+                    ->map(fn($rule) => static::compileRoutingRegex($rule)))
+            ->filter(fn($key, $matchers) => $matchers->count() > 0)
+            ->mapValues(fn($key, $val) => $val->toArray());
+
+        // Reduce the rules for each key into one rule, taking the first it's
+        // keyName and fieldAccessors. Note: If multiple fields can be configured
+        // to supply a value for the same keyName, this will not work. To date,
+        // we only support referring to one field with multiple parsing rules
+        // per header key.
+        $paramDescs = $rulesByKey->keys()
+            ->map(fn($key) => $rulesByKey->get($key, Vector::new())
+                ->reduce([], fn($val, $rule) => [
+                        'keyName' => $val['keyName'] ?? $key,
+                        'fieldAccessors' => $val['fieldAccessors'] ?? static::buildGetterChain($catalog, $msg, $rule->getField()),
+                ])
             );
+
+        // Include the list of segment matchers for a parsing rule.
+        $paramDescs = $paramDescs->map(fn($p) =>
+            isset($matchersByKey[$p['keyName']]) ? 
+                array_merge($p, ['matchers' => $matchersByKey[$p['keyName']]]) :
+                $p);
+        
+        return $paramDescs->toArray();
+    }
+
+    private static function hasMatcher(RoutingParameter $routingParam): bool
+    {
+        // Has a path_template in RoutingParameter and the path_template defines
+        // a segment matcher other than '**' ("match anything"), with the header
+        // key override.
+        $temp = $routingParam->getPathTemplate();
+        return !empty($temp) && str_contains($temp, '=') && !str_ends_with($temp, "=**}");
```

Comment:
str_contains / str_ends_with were introduced in PHP 8, if we'd like to keep compatibility with 7.4 we'll need to replace these.

Reply:
```suggestion
        return !empty($temp) && false !== strpos($temp, '=') && substr($temp, -4) !== '=**}';
```

Reply:
Ah good point. I've added some implementations with TODOs to remove them once we switch this repo to PHP 8.x. I also changed another spot that used these PHP8 methods.

Reply:
Ha @bshaffer we replied almost at the same time. I just added some implementations for these methods and replaced them so that read the same. WDYT?

Reply:
Your solution looks way better ;) nice work!